contracts/contracts/interfaces/staking-manager.md

```diff
@@ -0,0 +1,116 @@
+
+Once `RegisterSubnetValidatorTx` is accepted a warp message will be aggregated and relayed to this contract as a result of `RegisterSubnetValidatorTx`. Upon receiving and verifying this message (via `receiveRegisterValidator`), the contract should activate the validator and start accruing rewards. The message will only contain `messageID` which corresponds to the `messageID` in `registerValidator` request. The contract should verify these:
+
+- Verified warp message
+- The typecheck on the warp message (== ValidatorRegisteredMessage)
+- messageID is not empty
+- The message is not already used
+- (?) There is a pending request for this messageID
+- (?) The nodeID in the pending request is not a validator already
+- (?) Message is from P-Chain
+
+TODO: we might not want last 3 checks (with ?) to minimize reverts. The contract should be able to handle these cases without reverting. This is because the timing of the messages is not guaranteed and the contract should be able to handle out-of-order messages.
+
+If `RegisterSubnetValidatorTx` is not accepted, another warp message `InvalidValidatorRegisterMessage` will be relayed to the contract.
+
+## Remove Validator
+
+### removeValidator
+
+The manager contract can remove a validator gracefully by sending a warp message to P-Chain. `removeValidator` function should be called with the following inputs:
+
+- Subnet ID (bytes32): The ID of the subnet where the validator is being removed
+- Node ID (bytes32): The ID of the node that is being removed as a validator
+
+The contract should verify these:
+
+- The node is already registered as a validator
+- The address that registered the node is the same with the address that is removing the validator
+- (Optional) Validator being removed after minimum staking period
+
+After verification contract should perform the following:
+
+- Send a warp message to P-Chain for `SetSubnetValidatorWeightTx{messageID: relatedRegisterValidatorMessageID, nonce: MaxUint64, Weight: 0, }` transaction
+- Stop accruing rewards for the validator
+
+### receiveRemoveValidator
+
+Once the `SetSubnetValidatorWeightTx` is accepted, a warp message `InvalidValidatorRegisterMessage` will be relayed to the contract with `messageID` of the request. The message represents a messageID will not possibly become a valid validator. This essentialy means the message was not valid in the first place, or the validator has removed from the set. The InvalidValidatorRegisterMessage can either be a result of `removeValidator` from the contract or it can be originated directly from P-chain via `ExitValidatorSetTx`. In either case this ultimately tells the contract that a validator was evicted. See related ACP-77 section [here](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets#exitvalidatorsettx).
+
+The contract should verify these in `receiveRemoveValidator`:
+
+- Verified warp message
+- The typecheck on the warp message (== InvalidValidatorRegisterMessage)
```

Comment:
nit: I dislike the name of this message `InvalidValidatorRegisterMessage`, mainly because "Invalid....Message" seems weird as part of an expected flow.

Should we rename it to `ValidatorUnregistered` or `ValidatorNotRegistered` (or perhaps `Validation`)?
Not sure if it is too late for this

Reply:
yea i don't like it very much either. I'm not quite sure how to reflect "the message was invalid" + "the message won't be valid anymore. Maybe `DefunctMessage`?